### PR TITLE
Topic Banner

### PR DESCRIPTION
**Pull request by andymandias**
Initial implementation for #236.  Currently it builds on #221, but it can be disentangled if that complicates the process.

### CODE REVIEW

**Comment by andymandias** [comment]
Currently the topic bar searches the channel history (new to old) to find the topic, then puts it in a scrollable at the top of the channel pane.  If no topic is found, then no scrollable is created.  Included in the topic bar is the user that set it and when it was set, with the user's nick being able to be clicked on if they are in the channel.  I made it background-color text on an info color scrollable to keep it distinct from the other views, but picked the darkest shade of the info color to try to reduce how distracting it can be.  A maximum height can be set in the config to limit the height of the topic banner, with the default aiming to be two lines tall.

**Comment by casperstorm** [comment]
It's hard to review it since it builds on top of another PR. I would, if not too complicated, like to have it separate so we can work on them separately. Idea is great though, so I would like to see if we can land this feature.

**Comment by andymandias** [comment]
Absolutely, sorry about that.  Wasn't sure how GitHub would handle the PR, but I can see now it's a bit of a mess.  Will disentangle then force push.

**Comment by andymandias** [comment]
Disentangled :)

**Comment by andymandias** [comment]
Rebased and modified to take advantage of the serde code @tarkah provided in #221.

File: config.yaml
Comment (casperstorm) [inline config.yaml:90]: We should just name this `topic`

File: src/buffer/channel.rs
Comment (casperstorm) [inline src/buffer/channel.rs:116]: I feel like this part is quite complicated. We have nom inside view, and is doing quite a lot of calculations. We should tidy this up.

**Review by casperstorm — CHANGES_REQUESTED** [review]
Screenshot 2024-03-01 at 20 06 58

Also, i played a bit with the styling and think we can tone it a bit down from what you have currently.

File: config.yaml
Comment (andymandias) [reply config.yaml]: The more descriptive naming is a good idea.  Thanks & implemented :)  Wound up making the max height a struct field since I couldn't work out how to give it a default value otherwise.

File: config.yaml
Comment (andymandias) [reply config.yaml:97]: I briefly tried to have it set automatically, but I couldn't see how to get iced to give me the line height of the enclosed text (or to let me render offscreen so I can measure the line height).  I found the results where the max_height doesn't reveal an integer multiple of line height (plus padding) to be less than ideal.  I'll dig into it further.

File: config.yaml
Comment (andymandias) [reply config.yaml]: Definitely clearer.  Implemented :)

File: src/buffer/channel.rs
Comment (andymandias) [reply src/buffer/channel.rs]: Moved who/time line styling into its own function and removed nom for a much simpler split.  Had trouble finding documentation specifying the format for RPL_TOPICWHOTIME, but the simple split seems sufficient for our needs.

File: config.yaml
Comment (andymandias) [reply config.yaml]: Found `double_pass.rs` to use for calculating the rendered dimensions, such that the configuration can be in lines rather than requiring the user to know/determine the line height.  Might be too many passes to first find a window height percentage via `FillPortion` then the nearest number of lines though?

**Comment by casperstorm** [comment]
@tarkah mind taking a look at this as well?

**Comment by tarkah** [comment]
Let me look tomorrow, sorry for my delay on this

File: data/src/message/source.rs
Comment (tarkah) [inline data/src/message/source.rs:78]: try details.text.as_deref()

File: data/src/history/manager.rs
Comment (tarkah) [inline data/src/history/manager.rs:525]: I don't think we should rely on our history to extract the topic at all. 

When we join a channel, which happens every time we start up, we should always receive RPL_TOPIC from the server. I think we can just hold a map of latest topic messages per channel in our client state and update it when those messages come in.

Then we can can query the topic from our client state vs dealing with history.

We can then remove all the metadata from message we've added here except for a single new server kind for Topic so we can filter those messages out if the banner option is chosen.

I think this will simplify a lot. Let me know if you have any questions or if you'd prefer I implemented it, but if you want to give it a shot, let me know!

**Review by tarkah — CHANGES_REQUESTED** [review]

File: data/src/history/manager.rs
Comment (andymandias) [reply data/src/history/manager.rs:525]: Sounds like a good idea to me.  I will have to learn more about how commands/messages get passed around in iced/Halloy, but that's something I want to learn so I'd like to give it a shot.  I will likely have questions for you, but will see what I can figure out on my own first.

Comment (tarkah) [reply data/src/history/manager.rs:525]: Great! You can start here in data::client where we handle all messages:

https://github.com/squidowl/halloy/blob/cb6c624c77ef363df2068fa4c9002148b427dbe5/data/src/client.rs#L182

You can setup some new state on Client, update it when Topic messages come in, then expose a method to get the current topic for a server / channel and call it from view!

Comment (andymandias) [reply data/src/history/manager.rs]: This was a *very* helpful set of pointers, thank you!  I have made a first attempt at the recommended change.  I think it's quite a bit clearer/simpler and it seems to be working well here.

**Review by tarkah — APPROVED** [review]

File: data/src/history/manager.rs
Comment (tarkah) [reply data/src/history/manager.rs]: That is much better! Glad it was helpful.

I've pushed a few changes to get this across the finish line, let me know what you think.

@casperstorm LGTM from a technical standpoint, you can do a final design review

Comment (andymandias) [reply data/src/history/manager.rs]: Looks good to me.  Very instructive, thank you :)

The only "change" I have is a reversion of a change I made previously.  I had removed a `vertical_space` from before the `input_view` when switching to `spacing` because I couldn't figure out a clean way to include the same padding to the topic banner.  You found a clean way to do so for the topic banner and removed the `spacing` appropriately, but that meant there wasn't any spacing between the `input_view` and the `history_view` anymore.  So I put that back.

File: src/buffer/channel.rs
Comment (tarkah) [inline src/buffer/channel.rs:128]: Why are we adding this? The other buffers (Server and Query) don't add additional space here

**Review by tarkah — CHANGES_REQUESTED** [review]

File: src/buffer/channel.rs
Comment (tarkah) [reply src/buffer/channel.rs:128]: Edit: I just saw comment https://github.com/squidowl/halloy/pull/237#discussion_r1516813482, will reply there

File: data/src/history/manager.rs
Comment (tarkah) [reply data/src/history/manager.rs]: Can we add that same space to the other buffers then so we're consistent?

**Comment by casperstorm** [comment]
I am trying a few different designs tests, will report back tomorrow with some suggestions. Good job on this @andymandias.

File: data/src/history/manager.rs
Comment (tarkah) [reply data/src/history/manager.rs]: Man I'm sorry, total brain fart from me on this one.

**Review by tarkah — APPROVED** [review]

File: data/src/history/manager.rs
Comment (andymandias) [reply data/src/history/manager.rs]: Happens to the best of us :)

**Review by casperstorm — CHANGES_REQUESTED** [review]
Screenshot 2024-03-08 101704

I've played around with styling a bunch, and I think, for now, what I like the most is actually to keep it as simple as possible, just like the `nick_list` which uses: `theme::Scrollable::Hidden` as well as `theme::Text::Transparent` to blend it more with the pane header. That way we can also remove `theme::Scrollable::Banner` and `theme::Text::Banner`. When playing around with this, I found that changing the padding slightly at `src\buffer\channel.rs:160` worked the best:

```diff
    container(body)
        .width(Length::Fill)
        .height(Length::Fill)
-       .padding(8)
+       .padding([4, 8, 8, 8])
        .into()
```



Regarding layout i think we should go for a layout like this:
```
| topic   | users |
|---------|       |
| message |       |
```
Because it can look a little strange if topic isn't long enough and user list starts below:

image

Third, and last thing, I think we should add a pane button. Notice the button on the top image. Ive tried a bunch and kinda like the following:

```rust
pub fn topic<'a>() -> Text<'a> {
    to_text('\u{f5af}')
}
```
I think, with a button like this we can simplify the config a little as well and make it similar to Users:

```yaml
buffer:
    users:
      visible: true
```

Perhaps we could simply do:
```yaml
buffer:
    topic:
      visible: false
      max_lines: 2
```

Functionality should just be 1 to 1 with showing and hiding `nick_list`.
Let me know what you think about these suggetions, @andymandias.

**Comment by andymandias** [comment]
Thank you for the notes @casperstorm.

I think these are all good changes, and mesh into the Halloy style better.

Re: padding.  By removing the background color from the topic banner we can shrink the padding on it without any aesthetic changes.  Over here there's been a downside to the topic banner's current padding where it would show a sliver of the third line when only two lines were otherwise showing.  For example with the currently-committed padding I see:

Screenshot from 2024-03-08 14-32-56

So, unless there's an issue I'm going to remove the vertical padding from the topic banner.  I believe that will be equivalent to the padding change you've suggested in `channel.rs`, so I'm going to hold off on making that change.  I have no objection to changing the padding further, I just don't want to overshoot what you're suggesting.

I am slightly concerned that there's not a lot differentiating the topic from the rest of the channel text.  Would using something like a light horizontal rule fit within the Halloy style?

Screenshot from 2024-03-08 14-23-22

A seasoned user probably wouldn't need it, but a newbie might be able to use the extra affordance.

I like the pane button idea quite a bit.  I'm assuming toggling the topic button should show/hide the inline topic messages accordingly.  Is that right?  Or would it only toggle the visibility of the banner?

I think matching configuration options is ideal, but I am concerned it could be confusing to users if they set `visible: false` and they see inline topic messages.  Perhaps the options could be Inline, Banner, or false?

**Comment by casperstorm** [comment]
1. A divider between topic and content is all good to me. It looks good as you have it.
2. Re the sliver of the third line. Could a solution be adding a border and or background in the same color as the pane so it looks transparent but is colored? Just an idea, haven't tested if it actually works.
3. Re button. My initial thought was that config setting only affected banner, and button was a toggle for it. Similar to user list where you can set `visible: false`, but still quickly toggle to briefly check. What we could do is consider adding a `topic` variant to [server_messages](https://github.com/squidowl/halloy/wiki/Configuration#server-messages) which controls inline, and then rename `topic` under buffer to `topic_banner` with `visible: boolean, max_lines: u16`? That could solve all cases perhaps.
4. Re padding. I think we should aim to get the spacing between pane header, topic divider and from topic divider to content equal. See below image. Its a super small nit. 
image

**Comment by andymandias** [comment]
Thank you for the clarifications!  I'll make the changes, going with both banner and server messages configuration options as you describe.  No sweat if we wind up changing the configuration options again :)

**Comment by andymandias** [comment]
Some chores left to be done (e.g. example config update, code clean-up pass, etc), but I believe this has the new configuration and functionality we're looking for.  I struggled a little with the under-topic padding since its apparent height can be influenced by scroll state of the messages in the channel;  if I didn't get it right please feel free to fix (or let me know how far off I am).

**Review by casperstorm — APPROVED** [review]
I updated padding slightly, example config file, CHANGELOG and then rebased on main. We should update wiki once we merge.

I don't have anything else - this looks awesome, and what awesome feature. Do you have anything you want to clean up before we merge @andymandias?

**Comment by andymandias** [comment]
Did a quick review and didn't see anything I made too messy.  Looks good to merge to me.

Thanks for taking care of the rest of the chores @casperstorm, and for all the guidance from you and @tarkah.

**Pull request merged**
